### PR TITLE
feat: copy link button for search

### PR DESCRIPTION
**Pull request by patnir41**
Upon typing in the search box and hovering the mouse on the input box itself, a link button will appear for the user to copy the current URL and share their page view with others. When the user hovers away from the search box, the link button will disappear.

### CODE REVIEW

File: packages/api-explorer/src/components/SideNav/SideNav.tsx [hunk 1]
```diff
@@ -160,18 +165,33 @@ export const SideNav: FC<SideNavProps> = ({ headless = false, spec }) => {
...
+            isClearable
+          />
+          <CopyLinkButton
+            top={'24px'}
+            right={'58px'}
+            visible={showCopyLinkButton}
+          />
+        </Span>
+        <SearchMessage search={searchResults} />
+      </Box2>
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/SideNav/SideNav.tsx:194]: sidenav unit tests are missing

File: packages/api-explorer/src/components/CopyLinkButton/CopyLinkButton.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  )
+}
+
+const CopyLink = styled('span')<{
+  top: string
+  right: string
+  visible: boolean
+}>`
+  position: absolute;
+  top: ${({ top }) => top};
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/CopyLinkButton/CopyLinkButton.tsx:65]: would margins work here? 

in general, we should avoid custom spacing property values and leverage @looker/components built-in spacing values. See this [example](https://github.com/looker-open-source/sdk-codegen/blob/main/packages/api-explorer/src/components/SideNav/SideNavMethodTags.tsx#L54) on how to reference `theme` and available spacing props.

Also, I could see a use for having margin top/bottom/left/right. Consider supporting that now.

In the end, you should be able to use this component in ways similar to below:
```ts
<CopyLinkButton visible={true} mt="small" ml="large" />
```

**Review by josephaxisa — CHANGES_REQUESTED** [review]
looking good. Some feedback below

File: packages/api-explorer/src/components/CopyLinkButton/CopyLinkButton.tsx (hunk 2 above)
Comment (patnir41) [reply packages/api-explorer/src/components/CopyLinkButton/CopyLinkButton.tsx:65]: Currently blocked here, facing a problem where the copy button appears aligned to the far-left side and unable to be right-aligned.

**Review by jkaster — COMMENTED** [review]
Minor comment

File: packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * @param visible
+ */
+export const CopyLinkWrapper = ({
+  children,
+  visible = true,
+}: CopyLinkWrapperProps) => {
+  const [tooltipContent, setTooltipContent] = useState(COPY_TO_CLIPBOARD)
+  const [showCopyLinkButton, setShowCopyLinkButton] = useState(false)
+  const handleCopyLink = async () => {
+    // const adaptor = getEnvAdaptor()
```
Comment (jkaster) [inline packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx:52]: code cruft?

**Review by jkaster — COMMENTED** [review]
Haven't tested this yet, so just a comment about a commented out code line

File: packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+interface CopyLinkWrapperProps {
+  children: ReactNode | ReactNodeArray
+  visible?: boolean
+}
+
+const COPY_TO_CLIPBOARD = 'Copy to clipboard'
+
+/**
+ * Wrapper component that can render a hideable copy link button next to any
+ * passed in children components
```
Comment (josephaxisa) [inline packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx:40]:
```suggestion
 * Displays a copy link button on hover
 *
```

File: packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+ */
+import type { ReactNode, ReactNodeArray } from 'react'
+import React, { useState } from 'react'
+import { IconButton, Space } from '@looker/components'
+import { Link } from '@styled-icons/material-outlined/Link'
+
+interface CopyLinkWrapperProps {
+  children: ReactNode | ReactNodeArray
+  visible?: boolean
```
Comment (josephaxisa) [inline packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx:33]: documentation?

**Comment by jkaster** [comment]
FWIW this isn't working with the extension framework. Let me know if you need help setting up the development version of the extension with your local Looker instance.

File: packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ */
+export const CopyLinkWrapper = ({
+  children,
+  visible = true,
+}: CopyLinkWrapperProps) => {
+  const [tooltipContent, setTooltipContent] = useState(COPY_TO_CLIPBOARD)
+  const [showCopyLinkButton, setShowCopyLinkButton] = useState(false)
+  const handleCopyLink = async () => {
+    // const adaptor = getEnvAdaptor()
+    await navigator.clipboard.writeText(location.href)
```
Comment (patnir41) [inline packages/run-it/src/components/CopyLinkWrapper/CopyLinkWrapper.tsx:52]: TODO: set this to support browser vs extensions framework

File: packages/extension-utils/src/extensionAdaptor.ts [hunk 7]
```diff
@@ -59,6 +59,12 @@ export class ExtensionAdaptor
     )
   }
 
+  async copyToClipboard(location: { pathname: string; search: string }) {
```
Comment (patnir41) [inline packages/extension-utils/src/extensionAdaptor.ts:62]: Commenting to make visible to @josephaxisa @bryans99 : Per my discussion with Bryn today, I need a way to access the browser window.location object in this adaptor in order to append its pathname/search params to the hostname/url contained by the extensionSdk. I do this by passing in the `location` returned from `useLocation` when called in CopyLinkWrapper.

I experimented with a couple of different approaches for passing in the parameter, but settled on this typed object. This made it easier than dealing with the location object overall and its type issues when passing it in through the spec file as opposed to after getting it from the `useLocation` hook in CopyLinkWrapper. Thoughts on this approach?

Comment (bryans99) [reply packages/extension-utils/src/extensionAdaptor.ts:62]: Change to this. Route is available in lookerHostData. No need to pass it in
```

  async copyToClipboard() {
    const routeContext = this.extensionSdk.lookerHostData
    const { lookerHostData } = this.extensionSdk
    if (lookerHostData) {
      const { extensionId, route } = lookerHostData
      const url = `${routeContext?.hostUrl}/extensions/${extensionId}${route}`
      await this.extensionSdk.clipboardWrite(url)
    }
  }
```

Update `packages/extension-utils/src/adaptorUtils.ts` as follows

```
export interface IEnvironmentAdaptor extends IAuthAdaptor {
  /** Method for determining whether running in a browser or extension environment */
  isExtension(): boolean
  /** Method for retrieving a keyed value from local storage */
  localStorageGetItem(key: string): Promise<string | null>
  /** Method for setting a keyed value in local storage */
  localStorageSetItem(key: string, value: string): void
  /** Method for removing a keyed value from local storage */
  localStorageRemoveItem(key: string): void
  /** Theme settings for extension */
  themeOverrides(): ThemeOverrides
  /** Open a new browser window with the given url and target  */
  openBrowserWindow: (url: string, target?: string) => void
  /** error logger */
  logError: (error: Error, componentStack: string) => void
  /** copy to clipboard */
  copyToClipboard: () => Promise<void>
}
```

And CopyLinkWrapper.tsx as follows
```
  const handleCopyLink = async () => {
    await getEnvAdaptor().copyToClipboard()
    // await navigator.clipboard.writeText(location.href)
    setTooltipContent('Copied to clipboard')
  }
```

**Review by bryans99 — CHANGES_REQUESTED** [review]
See comments

**Review by bryans99 — COMMENTED** [review]
I've suggested cleaning up the extension copy method. 
1. handles possibility that fields are undefined (should not happen but typing says it could)
2. uses non deprecated field name
3. does not implicitly rename fields

**Review by bryans99 — COMMENTED** [review]
LGTM. 

Will let Jax give the final stamp

**Review by jkaster — COMMENTED** [review]
Looks like it's ready to test. I'll do that before EOD.

**Review by josephaxisa — APPROVED** [review]
lgtm. Thanks Niral!

**Review by jkaster — APPROVED** [review]
Works for me!

**Review by bryans99 — APPROVED** [review]

**Pull request merged**